### PR TITLE
Validator side of the collation protocol.

### PR DESCRIPTION
**Pull request by rphmeier**
Each collator collates for a specific parachain. A validator manages connected collators and assigns one of them for each parachain the "primary" role, while others are "backup". The primary is intended to push a collation to the validator whenever the validator is assigned to a specific parachain, while the backups do not.

This is implemented primarily with the `CollatorPool` struct, which is owned by the `PolkadotNetwork`. The `ConsensusNetwork` struct implements the `polkadot_consensus::Collators` trait faithfully now, by forwarding to the referenced `PolkadotNetwork` APIs.

Follow-up work includes:
  - saner key authentication protocol (#271) 
  - collator side of the network protocol
  - collator peer discovery
  - role reassignment based on performance

### CODE REVIEW

File: polkadot/network/src/lib.rs [hunk 1]
```diff
@@ -363,6 +371,8 @@ impl PolkadotProtocol {
 				send_polkadot_message(ctx, peer_id, Message::BlockData(req_id, block_data));
 			}
 			Message::BlockData(req_id, data) => self.on_block_data(ctx, peer_id, req_id, data),
+			Message::Collation(relay_parent, collation) => self.on_collation(ctx, peer_id, relay_parent, collation),
+			Message::CollatorRole(_) => {},
```
Comment (gnunicorn) [inline polkadot/network/src/lib.rs]: If I understand that Message right, this peer is informing us about our CollatorRole, no? And we just ignore it? Didn't it say up there in a TODO that we should get punished if we try to collate even if we aren't primary? Or is the implementation of this going to be part of the follow-up PR? Even in that case, maybe have a ToDo here, too?

File: polkadot/network/src/lib.rs [hunk 2]
```diff
@@ -483,6 +509,64 @@ impl Specialization<Block> for PolkadotProtocol {
...
+	fn on_collation(&mut self, ctx: &mut Context<Block>, from: PeerId, relay_parent: Hash, collation: Collation) {
+		let collation_para = collation.receipt.parachain_index;
+		let collated_acc = collation.receipt.collator;
+
+		match self.peers.get(&from) {
+			None => ctx.disconnect_peer(from),
+			Some(peer_info) => match peer_info.status.collating_for {
+				None => ctx.disable_peer(from),
+				Some((ref acc_id, ref para_id))
+					if para_id != &collation_para || acc_id != &collated_acc || collation.receipt.check_signature().is_err() => ctx.disable_peer(from),
```
Comment (gnunicorn) [inline polkadot/network/src/lib.rs]: That line is really too long to read ;) (150 chars)

File: polkadot/network/src/lib.rs [hunk 3]
```diff
@@ -483,6 +509,64 @@ impl Specialization<Block> for PolkadotProtocol {
...
+				Some((ref acc_id, _)) => self.collators.on_collation(acc_id.clone(), relay_parent, collation),
```
Comment (gnunicorn) [inline polkadot/network/src/lib.rs]: `on_collation` above states that _we must have checked the signature_ prior, and although this happens in the match check before, arriving here, that hasn't necessarily happened, does it? Also relying on a check to have happened in the match prior sounds like something that could easily be missed when changing this later. I'd prefer to have a more explicit guard around it, that makes sure check_signature has been called before on_collation regardless. Something like 
```rust
Some((ref acc_id, ref para_id)) => {
  if para_id == &collation_para && acc_id == &collated_acc && collation.receipt.check_signature().is_ok() {
      self.collators.on_collation(acc_id.clone(), relay_parent, collation)
  } else {
    ctx.disable_peer(from)
  }
}
```

File: polkadot/network/src/lib.rs [hunk 4]
```diff
@@ -483,6 +509,64 @@ impl Specialization<Block> for PolkadotProtocol {
...
+		let (tx, rx) = oneshot::channel();
+		self.collators.await_collation(relay_parent, para_id, tx);
+		rx
+	}
+
+	// get connected peer with given account ID for collation.
+	fn collator_peer_id(&self, account_id: AccountId) -> Option<PeerId> {
+		self.peers
+			.iter()
+			.filter(|&(_, info)| info.status.collating_for.as_ref().map_or(false, |&(ref acc_id, _)| acc_id == &account_id))
```
Comment (gnunicorn) [inline polkadot/network/src/lib.rs]: 125 chars. please wrap ;) .

File: polkadot/primitives/src/parachain.rs [hunk 5]
```diff
@@ -195,6 +195,17 @@ impl CandidateReceipt {
 		use runtime_primitives::traits::{BlakeTwo256, Hash};
 		BlakeTwo256::hash_of(self)
 	}
+
+	/// Check integrity vs. provided block data.
+	pub fn check_signature(&self) -> Result<(), ()> {
+		use runtime_primitives::traits::Verify;
+
+		if self.signature.verify(&self.signature.0[..], &self.collator) {
```
Comment (pepyakin) [inline polkadot/primitives/src/parachain.rs]: Shouldn't this take `block_data_hash`?

File: polkadot/network/src/collator_pool.rs [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+		CollatorPool {
+			collators: HashMap::new(),
+			parachain_collators: HashMap::new(),
+			collations: HashMap::new(),
+		}
+	}
+
+	/// Call when a new collator is authenticated. Returns the role.
+	pub fn on_new_collator(&mut self, account_id: AccountId, para_id: ParaId) -> Role {
+		self.collators.insert(account_id.clone(), para_id);
```
[the remark this answers is not in the document]
Comment (rphmeier) [reply polkadot/network/src/collator_pool.rs]: #271 will ensure that we are connected to only one peer ID with a claimed `AccountId` at a time, and each `AccountId` can only collate for one parachain during the span of the connection in the current protocol.

File: polkadot/network/src/lib.rs (hunk 1 above)
Comment (rphmeier) [reply polkadot/network/src/lib.rs]: that's the follow-up PR. Can add a TODO but it'll follow-up really shortly so I don't know for whose benefit that is.

File: polkadot/network/src/lib.rs (hunk 3 above)
Comment (rphmeier) [reply polkadot/network/src/lib.rs]: have rewritten to this before seeing this comment. Agree

**Comment by gavofyork** [comment]
> A validator ... assigns one of [the connected collators] for each parachain the "primary" role, while others are "backup". The primary is intended to push a collation to the validator whenever the validator is assigned to a specific parachain, while the backups do not.

this doesn't sound like The Plan. why the big change? validators should be listening out for candidate blocks on their current parachain generally and figuring out the best candidate with some metric (tbd). they shouldn't be picking a winning collator up front.

**Comment by rphmeier** [comment]
> validators should be listening out for candidate blocks on their current parachain generally and figuring out the best candidate with some metric (tbd). they shouldn't be picking a winning collator up front.

this refers to the collations put forth by validators, not the process by which validators choose which collation to put forth into the consensus.

if each validator accepts block proposals from all connected collators it will probably use up too much network bandwidth. I'm not particularly tied to the primary/backup scheme, it was just the simplest thing to implement for PoC-2 other than accepting collations from all collators. But we can also change the primary periodically or have multiple primaries at a time.

**Comment by gavofyork** [comment]
so for poc-2 this is fine. but eventually (poc-5/6), i'd like to see a time-limit-based approach and as many PoV candidates evaluated as possible within it, ideally parallelised over cores. some of the game theory (such as golden tickets for collators to encourage collator-rotation) doesn't work unless validators evaluate multiple candidates.

**Comment by gavofyork** [comment]
https://github.com/paritytech/polkadot/issues/301

**Pull request merged**